core/src/main/java/org/mule/execution/AsyncResponseFlowProcessingPhase.java

```diff
@@ -60,12 +65,46 @@ public MuleEvent process() throws Exception
                     {
                         MuleEvent muleEvent = template.getMuleEvent();
                         fireNotification(muleEvent, MESSAGE_RECEIVED);
-                        muleEvent = template.routeEvent(muleEvent);
-                        return muleEvent;
+                        if (muleEvent.getExchangePattern().hasResponse() && !muleEvent.isSynchronous())
```

Comment:
Sure.  I could alternatively add a isNonBlocking() method to MuleEvent, because I use this else where. WDYT?